### DISCUSSION TITLE
Enhance the data prop of the upload component

### DISCUSSION THREAD

**Discussion by jianjunyuu** [element-plus/element-plus#14016] [Feature Request]
拓展Upload组件的data属性，支持可异步返回data的能力

```typescript
type UploadData = Mutable<Record<string, any>>
data: Awaitable<UploadData> | ((rawFile: UploadRawFile) => Awaitable<UploadData>)
```

因为
我们知道，目前的upload组件的data属性的类型是Record<string, any>
当我们批量上传多个文件时候，会使用同一份data

如果想要针对每个文件的上传，自定义上传的data，那么只能在`beforeUpload`中去修改data
但这最终每个文件上传的data永远只能是最后一次修改的状态，因为`doUpload`是异步执行的，而修改data的操作都在同步中完成
所以有了 https://github.com/element-plus/element-plus/pull/12029, 可以在`beforeUpload`前就把每次修改的data给缓存了下来，保证了每次上传的data都是当前那一刻的

可这还是不够完美，比如：[想要异步的操作data](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGVsLXVwbG9hZFxuICAgIHYtbW9kZWw6ZmlsZS1saXN0PVwiZmlsZUxpc3RcIlxuICAgIGFjdGlvbj1cImh0dHBzOi8vb3NzLWNuLWJlaWppbmcuYWxpeXVuY3MuY29tXCJcbiAgICA6ZGF0YT1cInVwbG9hZERhdGFcIlxuICAgIDpiZWZvcmUtdXBsb2FkPVwiYmVmb3JlVXBsb2FkXCJcbiAgICA6b24tc3VjY2Vzcz1cInVwbG9hZFN1Y2Nlc3NcIlxuICAgIDpsaW1pdD1cIjk5OTlcIlxuICA+XG4gICAgPGVsLWJ1dHRvbiB0eXBlPVwicHJpbWFyeVwiPkNsaWNrIHRvIHVwbG9hZDwvZWwtYnV0dG9uPlxuICAgIDx0ZW1wbGF0ZSAjdGlwPlxuICAgICAgPGRpdiBjbGFzcz1cImVsLXVwbG9hZF9fdGlwXCI+XG4gICAgICAgIGpwZy9wbmcgZmlsZXMgd2l0aCBhIHNpemUgbGVzcyB0aGFuIDUwMEtCLlxuICAgICAgPC9kaXY+XG4gICAgPC90ZW1wbGF0ZT5cbiAgPC9lbC11cGxvYWQ+XG48L3RlbXBsYXRlPlxuPHNjcmlwdCBsYW5nPVwidHNcIiBzZXR1cD5cbmltcG9ydCB7IHJlZiwgcmVhY3RpdmUgfSBmcm9tICd2dWUnXG5cbmltcG9ydCB0eXBlIHsgVXBsb2FkUHJvcHMsIFVwbG9hZFVzZXJGaWxlIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuXG5jb25zdCBmaWxlTGlzdCA9IHJlZjxVcGxvYWRVc2VyRmlsZVtdPihbXG5dKVxuXG5jb25zdCB1cGxvYWREYXRhID0gcmVhY3RpdmU8e1xuICBrZXk6IHN0cmluZztcbn0+KHtcbiAga2V5OiAnMTExMTEnLFxufSk7XG5cbmNvbnN0IGJlZm9yZVVwbG9hZDogVXBsb2FkUHJvcHNbJ2JlZm9yZVVwbG9hZCddID0gKHJhd0ZpbGUpID0+IHtcbiAgLy8g5LuOMi4zLjPniYjmnKzlvIDlp4vvvIxkYXRh5Lit55qE5pWw5o2u5aaC5p6c5piv5byC5q2l5pu05pS555qE77yM5ZyocmVzb2x2ZSh0cnVlKei/m+ihjOS4iuS8oOaXtu+8jOacgOaWsOeahGRhdGHkuK3nmoTmlbDmja7ml6Dms5Xooqvmm7TmlrBcbiAgLy8gMi4zLjPniYjmnKzku6XkuIvliJnkuI3lj5flvbHlk41cbiAgcmV0dXJuIG5ldyBQcm9taXNlKHJlc29sdmUgPT4ge1xuICAgIC8vIOaooeaLn+W8guatpeWkhOeQhlxuICAgIC8vIOivt+axgmFwaeaLv29zc+eahOmFjee9ruaVsOaNrlxuICAgIHNldFRpbWVvdXQoKCkgPT4ge1xuICAgICAgdXBsb2FkRGF0YS5rZXkgPSAnMjIyMic7XG4gICAgICAvLyDms6jmhI/nnIvmjqfliLblj7BuZXR3b3Jr5LiLb3Nz5o6l5Y+j5Y+C5pWw77yMa2V55a2X5q615bqU6K+l5Li6MjIyMlxuICAgICAgcmVzb2x2ZSh0cnVlKVxuICAgIH0sIDEwMDApXG4gIH0pO1xuICBcbn07XG5cbmNvbnN0IHVwbG9hZFN1Y2Nlc3MgPSAoKSA9PiB7XG4gIGNvbnNvbGUubG9nKCd1cGxvYWQtc3VjY2VzcycpXG59XG48L3NjcmlwdD5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZSxcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn1cbiIsIl9vIjp7fX0=)
就无法满足了。

基于以上的场景，为了丰富data能够支持的复杂场景
Related pr https://github.com/element-plus/element-plus/pull/14015

**Comment by btea** [comment]
What do you think about @HeftyKoo's [point of view](https://github.com/element-plus/element-plus/issues/13925#issuecomment-1681581284)?  @tolking @chenxch @ryuhangyeong

**Reply by GenerQAQ** [reply]
Sorry for being too busy during this period of time. I tried to solve the problem after reproducing it before, and found that if the promises in beforeUpload still have asynchronous methods, there will be problems. I haven't found a good solution yet. Please help me too~ @chenxch
